File: frontend/src/model/video-playback-coordinator.ts
type PlaybackListener = (activeId: string | null) => void;

const listeners = new Set<PlaybackListener>();
let activeId: string | null = null;

/** 再生中の動画。複数あっても実際に音と映像を出すのは常に1つにする。 */
export function currentVideoPlayback(): string | null {
  return activeId;
}

export function claimVideoPlayback(id: string): void {
  if (activeId === id) return;
  activeId = id;
  listeners.forEach((listener) => {
    listener(activeId);
  });
}

export function releaseVideoPlayback(id: string): void {
  if (activeId !== id) return;
  activeId = null;
  listeners.forEach((listener) => {
    listener(activeId);
  });
}

export function subscribeVideoPlayback(listener: PlaybackListener): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/** テスト用の初期化。製品コードからは呼ばない。 */
export function resetVideoPlaybackForTesting(): void {
  activeId = null;
  listeners.clear();
}

File: android/app/src/main/java/dev/nytweetdeck/android/ui/VideoPlaybackCoordinator.kt
package dev.nytweetdeck.android.ui

import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicReference

/**
 * 同時に再生される動画を常に1つに保つ。再生要素は複数あってよいが、
 * 実際に音と映像を出すのは最後に再生を始めた1つだけにする。
 */
object VideoPlaybackCoordinator {
    private val activeId = AtomicReference<String?>(null)
    private val listeners = CopyOnWriteArrayList<(String?) -> Unit>()

    fun current(): String? = activeId.get()

    fun claim(id: String) {
        if (activeId.getAndSet(id) == id) return
        listeners.forEach { it(id) }
    }

    fun release(id: String) {
        if (activeId.compareAndSet(id, null)) {
            listeners.forEach { it(null) }
        }
    }

    fun observe(listener: (String?) -> Unit): AutoCloseable {
        listeners.add(listener)
        return AutoCloseable { listeners.remove(listener) }
    }

    internal fun resetForTesting() {
        activeId.set(null)
        listeners.clear()
    }
}
